solarwinds_apm/oboe/oboe_sampler.py

```diff
@@ -0,0 +1,702 @@
+# © 2025 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+from __future__ import annotations
+
+import logging
+import re
+import time
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from enum import IntEnum, auto
+
+from opentelemetry.context import Context
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.trace.sampling import (
+    Attributes,
+    Decision,
+    Sampler,
+    SamplingResult,
+)
+from opentelemetry.trace import Link, SpanKind, TraceFlags, get_current_span
+from opentelemetry.trace.span import Span, TraceState
+from typing_extensions import override
+
+from solarwinds_apm.oboe.dice import _Dice
+from solarwinds_apm.oboe.metrics import Counters
+from solarwinds_apm.oboe.settings import (
+    BucketType,
+    Flags,
+    LocalSettings,
+    Settings,
+    merge,
+)
+from solarwinds_apm.oboe.token_bucket import _TokenBucket
+from solarwinds_apm.oboe.trace_options import (
+    Auth,
+    RequestHeaders,
+    ResponseHeaders,
+    TraceOptionsResponse,
+    TraceOptionsWithResponse,
+    TriggerTrace,
+    parse_trace_options,
+    stringify_trace_options_response,
+    validate_signature,
+)
+
+TRACESTATE_REGEXP = r"^[0-9a-f]{16}-[0-9a-f]{2}$"
+DICE_SCALE = 1_000_000
+
+SW_KEYS_ATTRIBUTE = "SWKeys"
+PARENT_ID_ATTRIBUTE = "sw.tracestate_parent_id"
+SAMPLE_RATE_ATTRIBUTE = "SampleRate"
+SAMPLE_SOURCE_ATTRIBUTE = "SampleSource"
+BUCKET_CAPACITY_ATTRIBUTE = "BucketCapacity"
+BUCKET_RATE_ATTRIBUTE = "BucketRate"
+TRIGGERED_TRACE_ATTRIBUTE = "TriggeredTrace"
+
```

Comment:
I _think_ all attributes the new Sampler needs to set are accounted for. ✔️ The old sampler [sets an additional attribute](https://github.com/solarwinds/apm-python/blob/4caf5852935f3b32bdccf0bf1206f5132bd16291/solarwinds_apm/sampler.py#L552-L557) `sw.w3c.tracestate` but I think that's related to inbound metrics i.e. [this old PR](https://github.com/solarwinds/apm-python/pull/23). @cheempz is this correct?

Reply:
Added in sampler